### PR TITLE
gem CondFormats

### PR DESCRIPTION
**Pull request by jshlee**
splitting PR #21259

 - CondFormats for GEM and ME0
 - electronics mapping for gem amd me0
 - gem also has Dead and masked Strips

@slava77 @kpedro88 @hyunyong

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: CondFormats/GEMObjects/interface/GEMMapItem.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    std::string chamberLabel;
+    int chamberId;
+    int endcap;
+    int station;
+    int ring;
+    int chamber;
+    int gemIndex;
+    int layerIndex;
+    int stripIndex;
+    /* int anodeIndex; */
```
Comment (kpedro88) [inline CondFormats/GEMObjects/interface/GEMMapItem.h]: remove commented-out code

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+const std::string & GEMEMap::version() const{
+  return theVersion;
+}
+
+GEMROmap* GEMEMap::convert() const{
+  GEMROmap* romap=new GEMROmap();
+
+  std::vector<GEMEMap::GEMVFatMaptype>::const_iterator imap;
+  for (imap=this->theVFatMaptype.begin(); imap<this->theVFatMaptype.end();imap++){
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: range-based for loop

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+GEMROmap* GEMEMap::convertDummy() const{
+  GEMROmap* romap=new GEMROmap();
+  uint16_t amcId = 1; //amc
+  uint16_t gebId = 1; 
+  romap->addAMC(amcId);
+	
+  for (int re = -1; re <= 1; re = re+2) {
+    for (int st = 1; st<=2; ++st) {
+      // for (int ch = GEMDetId::minChamberId; ch<=GEMDetId::maxChamberId; ++ch) {
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: remove commented-out code

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      // 	for (int ly = 1; ly<=GEMDetId::maxLayerId; ++ly) {
+      for (int ch = 1; ch<=36; ++ch) {
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: There are a lot of magic numbers in this code. Please make them into descriptively-named constants.

File: CondFormats/GEMObjects/src/ME0EMap.cc [hunk 5]
```diff
@@ -0,0 +1,96 @@
+#include <iostream>
+#include "CondFormats/GEMObjects/interface/ME0EMap.h"
+#include "CondFormats/GEMObjects/interface/ME0ROmap.h"
+
+#include <DataFormats/MuonDetId/interface/ME0DetId.h>
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/ME0EMap.cc]: use quotes rather than <>

File: CondFormats/GEMObjects/src/ME0EMap.cc [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+const std::string & ME0EMap::version() const{
+  return theVersion;
+}
+
+ME0ROmap* ME0EMap::convert() const{
+  ME0ROmap* romap=new ME0ROmap();
+
+  std::vector<ME0EMap::ME0VFatMaptype>::const_iterator imap;
+  for (imap=this->theVFatMaptype.begin(); imap<this->theVFatMaptype.end();imap++){
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/ME0EMap.cc]: range-based for loop

File: CondFormats/GEMObjects/src/ME0EMap.cc [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  uint16_t gebId = 1; 
+  romap->addAMC(amcId);
+	
+  for (int re = -1; re <= 1; re = re+2) {
+    for (int ch = ME0DetId::minChamberId; ch<=ME0DetId::maxChamberId; ++ch) {
+      for (int ly = 1; ly<=ME0DetId::maxLayerId; ++ly) {
+	  
+	// 1 geb per chamber
+	// 24 gebs per amc
+	if (gebId > 25){
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/ME0EMap.cc]: same comment about magic numbers

**Comment by kpedro88** [comment]
assign upgrade

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	    
+	    for (int nVfat = 0; nVfat < maxVFat; ++nVfat){
+	      chipId++;
+	      
+	      for (unsigned chan = 0; chan < 128; ++chan){	    
+		GEMROmap::dCoord dc;
+		dc.stripId = ++stripId;
+		dc.gemDetId = gemId;
+
+		uint32_t vfatId = (amcId << 17) | (gebId << 12) | chipId;
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: still a bunch of magic numbers like these ones... (from experience, it is much easier to maintain if bitmasks, etc. are labeled constants)

**Comment by kpedro88** [comment]
please test

File: CondFormats/GEMObjects/src/GEMEMap.cc (hunk 8 above)
Comment (jshlee) [reply CondFormats/GEMObjects/src/GEMEMap.cc]: done!

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  return theVersion;
+}
+
+GEMROmap* GEMEMap::convert() const{
+  GEMROmap* romap=new GEMROmap();
+
+  for (auto imap : theVFatMaptype){
+    for (unsigned int ix=0;ix<imap.strip_number.size();ix++){
+      GEMROmap::eCoord ec;
+      ec.vfatId= imap.vfatId[ix] - 0xf000;// chip ID is 12 bits
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: magic number

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+      ec.channelId=imap.vfat_chnnel_number[ix];
+      GEMROmap::dCoord dc;
+      dc.stripId = 1 + imap.strip_number[ix]+(imap.iPhi[ix]-1)%3*128;
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: several magic numbers

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+	
+  for (int re = -1; re <= 1; re = re+2) {
+    for (int st = GEMDetId::minStationId; st<=GEMDetId::maxStationId; ++st) {
+      for (int ch = 1; ch<=GEMDetId::maxChamberId; ++ch) {
+      	for (int ly = 1; ly<=GEMDetId::maxLayerId; ++ly) {
+	  
+	  // 1 geb per chamber
+	  // 24 gebs per amc
+	  // make new amcId once 24 gebs are used up
+	  if (gebId > 25){
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: magic number

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+	    gebId = 1;
+	    amcId++;
+	    romap->addAMC(amcId);
+	  }
+
+	  romap->addAMC2GEB(amcId, gebId);
+	  
+	  GEMDetId chamDetId(re, 1, st, ly, ch, 0);
+	  uint32_t chamberId = (amcId << 5) | gebId;
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: magic number

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+	  romap->addAMC2GEB(amcId, gebId);
+	  
+	  GEMDetId chamDetId(re, 1, st, ly, ch, 0);
+	  uint32_t chamberId = (amcId << 5) | gebId;	  
+	  romap->add(chamDetId,chamberId);
+	  romap->add(chamberId,chamDetId);
+	  
+	  uint16_t chipId = 0;	 	  
+	  for (int roll = 1; roll<=GEMDetId::maxRollId; ++roll) {
+	    int maxVFat = 3;
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: magic number?
(also, if this is only based on the station, it could be defined in the outer loop)

File: CondFormats/GEMObjects/src/GEMEMap.cc [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+	      maxVFat = 6;
+	    }
+	    
+	    int stripId = 0;
+	    GEMDetId gemId(re, 1, st, ly, ch, roll);
+	    
+	    for (int nVfat = 0; nVfat < maxVFat; ++nVfat){
+	      chipId++;
+	      
+	      for (unsigned chan = 0; chan < 128; ++chan){
```
Comment (kpedro88) [inline CondFormats/GEMObjects/src/GEMEMap.cc]: magic number

**Review by kpedro88 — COMMENTED** [review]
@jshlee can you do the same for the rest of the GEM magic numbers I've pointed out (as well as the corresponding ME0 ones)?

**Comment by kpedro88** [comment]
please test
thanks @jshlee

**Comment by ggovi** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by arunhep** [comment]
+1

File: CondFormats/GEMObjects/interface/GEMMapItem.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
+#ifndef GEMMapItem_h
```
Comment (davidlange6) [inline CondFormats/GEMObjects/interface/GEMMapItem.h]: hi @jshlee -please use ifndef <subsystem>_<package>_<file>_h here and elsewhere for include guards

File: CondFormats/GEMObjects/interface/GEMMaskedStrips.h [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    int rawId;
+    int strip;
+    COND_SERIALIZABLE;
+  };
+  
+  GEMMaskedStrips(){}
+  ~GEMMaskedStrips(){}
+
+  std::vector<MaskItem> const & getMaskVec() const {return MaskVec;}
+  std::vector<MaskItem> MaskVec;
```
Comment (davidlange6) [inline CondFormats/GEMObjects/interface/GEMMaskedStrips.h]: please name this maskVec_ and make it private

File: CondFormats/GEMObjects/src/classes.h [hunk 17]
```diff
@@ -0,0 +1,8 @@
+#include "CondFormats/GEMObjects/interface/GEMEMap.h"
```
Comment (davidlange6) [inline CondFormats/GEMObjects/src/classes.h]: is this file needed?

File: CondFormats/GEMObjects/interface/GEMMaskedStrips.h (hunk 16 above)
Comment (jshlee) [reply CondFormats/GEMObjects/interface/GEMMaskedStrips.h]: done

File: CondFormats/GEMObjects/interface/GEMMapItem.h (hunk 15 above)
Comment (jshlee) [reply CondFormats/GEMObjects/interface/GEMMapItem.h]: hi @davidlange6 - I dont quite follow what you mean by the guards. The headers already all have ifndef.

File: CondFormats/GEMObjects/src/classes.h (hunk 17 above)
Comment (jshlee) [reply CondFormats/GEMObjects/src/classes.h]: yes, i think so. Does not compile without this file

Comment (andjuo) [reply CondFormats/GEMObjects/src/classes.h]: From CondObjectsTutorial:
"We also need a src/headers.h file, which holds include statements to all our serializable classes. This file is read by the serialization system to find the serializable classes and generate serialization code for them."

**Comment by davidlange6** [comment]
please test

**Comment by perrotta** [comment]
ping: can this proceed so that we can go on with the integration of the GEM packer/unpacker?
Thank you

**Comment by kpedro88** [comment]
+1

**Comment by arunhep** [comment]
+1

**Comment by arunhep** [comment]
this PR can be merged as the changes last happened are not really related to DB. And @ggovi has signed once already.

**Comment by arunhep** [comment]
@davidlange6 can this be merged please?

**Comment by slava77** [comment]
@ggovi 
your signature is still missing here 
OTOH, @davidlange6  this has been signed by db earlier before the cleanup commits. So, could this  perhaps be merged without waiting?

File: CondFormats/GEMObjects/interface/GEMMapItem.h (hunk 15 above)
Comment (davidlange6) [reply CondFormats/GEMObjects/interface/GEMMapItem.h]: sorry - missed the question (and indeed, my comment makes no sense for some reason). The include guards should include not only the name but the subsystem and package name by convention. Eg, CondFormats_GEMObjects_GEMMapItem_h

**Comment by jshlee** [comment]
@davidlange6 - added full guard names

**Comment by kpedro88** [comment]
please test

**Comment by arunhep** [comment]
+1
provided jenkins agree

**Comment by davidlange6** [comment]
merge

**Comment by davidlange6** [comment]
thanks!
